File: app/src/main/java/com/osrsd/cache/provider/ObjectProvider.java
package com.osrsd.cache.provider;

import com.osrsd.cache.def.ObjectDefinition;
import com.osrsd.cache.util.ByteBufferExt;
import com.osrsd.spring.domain.Definition;
import lombok.extern.slf4j.Slf4j;

import java.nio.ByteBuffer;
import java.util.stream.IntStream;

@Slf4j
public final class ObjectProvider {

    public static Definition decode(ByteBuffer buffer, ObjectDefinition definition) {
        do {
            int opcode = buffer.get() & 0xff;
            if (opcode == 0) {
                break;
            }

            if (opcode == 1) {
                int length = buffer.get() & 0xff;
                if (length > 0) {
                    int[] objectTypes = new int[length];
                    int[] objectModels = new int[length];
                    IntStream.range(0, length).forEach(index -> {
                        objectModels[index] = buffer.getShort() & 0xffff;
                        objectTypes[index] = buffer.get() & 0xff;
                    });
                    definition.setObjectTypes(objectTypes);
                    definition.setObjectModels(objectModels);
                }
            } else if (opcode == 2) {
                definition.setName(ByteBufferExt.INSTANCE.getString(buffer));
            } else if (opcode == 5) {
                int length = buffer.get() & 0xff;
                if (length > 0) {
                    definition.setObjectTypes(null);
                    definition.setObjectModels(IntStream.range(0, length).map(index -> buffer.getShort() & 0xffff).toArray());
                }
            } else if (opcode == 14) {
                definition.setSizeX(buffer.get() & 0xff);
            } else if (opcode == 15) {
                definition.setSizeY(buffer.get() & 0xff);
            } else if (opcode == 17) {
                definition.setInteractType(0);
                definition.setBlocksProjectile(false);
            } else if (opcode == 18) {
                definition.setBlocksProjectile(false);
            } else if (opcode == 19) {
                definition.setWallOrDoor(buffer.get() & 0xff);
            } else if (opcode == 21) {
                definition.setContouredGround(0);
            } else if (opcode == 22) {
                definition.setMergeNormals(true);
            } else if (opcode == 23) {
                definition.setABool2111(true);
            } else if (opcode == 24) {
                definition.setAnimationID(buffer.getShort() & 0xffff);
                if (definition.getAnimationID() == 0xffff) {
                    definition.setAnimationID(-1);
                }
            } else if (opcode == 27) {
                definition.setInteractType(1);
            } else if (opcode == 28) {
                definition.setDecorDisplacement(buffer.get() & 0xff);
            } else if (opcode == 29) {
                definition.setAmbient(buffer.get());
            } else if (opcode == 39) {
                definition.setContrast(buffer.get() * 25);
            } else if (opcode >= 30 && opcode < 35) {
                String[] actions = definition.getActions();
                actions[opcode - 30] = ByteBufferExt.INSTANCE.getString(buffer);
                if (actions[opcode - 30].equalsIgnoreCase("Hidden")) {
                    actions[opcode - 30] = null;
                }
            } else if (opcode == 40) {
                int length = buffer.get() & 0xff;
                short[] recolorToFind = new short[length];
                short[] recolorToReplace = new short[length];
                IntStream.range(0, length).forEach(index -> {
                    recolorToFind[index] = buffer.getShort();
                    recolorToReplace[index] = buffer.getShort();
                });
                definition.setRecolorToFind(recolorToFind);
                definition.setRecolorToReplace(recolorToReplace);
            } else if (opcode == 41) {
                int length = buffer.get() & 0xff;
                short[] retextureToFind = new short[length];
                short[] textureToReplace = new short[length];
                IntStream.range(0, length).forEach(index -> {
                    retextureToFind[index] = buffer.getShort();
                    textureToReplace[index] = buffer.getShort();
                });
                definition.setRetextureToFind(retextureToFind);
                definition.setTextureToReplace(textureToReplace);
            } else if (opcode == 62) {
                definition.setRotated(true);
            } else if (opcode == 64) {
                definition.setShadow(false);
            } else if (opcode == 65) {
                definition.setModelSizeX(buffer.getShort() & 0xffff);
            } else if (opcode == 66) {
                definition.setModelSizeHeight(buffer.getShort() & 0xffff);
            } else if (opcode == 67) {
                definition.setModelSizeY(buffer.getShort() & 0xffff);
            } else if (opcode == 68) {
                definition.setMapSceneID(buffer.getShort() & 0xffff);
            } else if (opcode == 69) {
                definition.setBlockingMask(buffer.get());
            } else if (opcode == 70) {
                definition.setOffsetX(buffer.getShort() & 0xffff);
            } else if (opcode == 71) {
                definition.setOffsetHeight(buffer.getShort() & 0xffff);
            } else if (opcode == 72) {
                definition.setOffsetY(buffer.getShort() & 0xffff);
            } else if (opcode == 73) {
                definition.setObstructsGround(true);
            } else if (opcode == 74) {
                definition.setHollow(true);
            } else if (opcode == 75) {
                definition.setSupportsItems(buffer.get() & 0xff);
            } else if (opcode == 77) {
                int varpID = buffer.getShort() & 0xffff;
                if (varpID == 0xFFFF) {
                    varpID = -1;
                }
                definition.setVarbitID(varpID);
                int configId = buffer.getShort() & 0xffff;
                if (configId == 0xFFFF) {
                    configId = -1;
                }
                definition.setVarpID(configId);
                int length = buffer.get() & 0xff;
                int[] configChangeDest = new int[length + 2];
                IntStream.rangeClosed(0, length).forEach(index -> {
                    configChangeDest[index] = buffer.getShort() & 0xffff;
                    if (0xffff == configChangeDest[index]) {
                        configChangeDest[index] = -1;
                    }
                });
                configChangeDest[length + 1] = -1;
                definition.setConfigChangeDest(configChangeDest);
            } else if (opcode == 78) {
                definition.setAmbientSoundId(buffer.getShort() & 0xffff);
                definition.setAnInt2083(buffer.get() & 0xff);
            } else if (opcode == 79) {
                definition.setAnInt2112(buffer.getShort() & 0xffff);
                definition.setAnInt2113(buffer.getShort() & 0xffff);
                definition.setAnInt2083(buffer.get() & 0xff);
                int length = buffer.get() & 0xff;
                definition.setAnIntArray2084(IntStream.range(0, length).map(index -> buffer.getShort() & 0xffff).toArray());
            } else if (opcode == 81) {
                definition.setContouredGround((buffer.get() & 0xff) * 256);
            } else if (opcode == 82) {
                definition.setMapAreaId(buffer.getShort() & 0xffff);
            } else if (opcode == 92) {
                int varpID = buffer.getShort() & 0xffff;
                if (varpID == 0xffff) {
                    varpID = -1;
                }
                definition.setVarbitID(varpID);
                int configId = buffer.getShort() & 0xffff;
                if (configId == 0xffff) {
                    configId = -1;
                }
                definition.setVarpID(configId);
                int var = buffer.getShort() & 0xffff;
                if (var == 0xffff) {
                    var = -1;
                }
                int length = buffer.get() & 0xff;
                int[] configChangeDest = new int[length + 2];
                IntStream.rangeClosed(0, length).forEach(index -> {
                    configChangeDest[index] = buffer.getShort() & 0xffff;
                    if (0xffff == configChangeDest[index]) {
                        configChangeDest[index] = -1;
                    }
                });
                configChangeDest[length + 1] = var;
                definition.setConfigChangeDest(configChangeDest);
            } else if (opcode == 249) {
                int length = buffer.get() & 0xff;
                for (int i = 0; i < length; i++) {
                    boolean isString = (buffer.get() & 0xff) == 1;
                    int key = ByteBufferExt.INSTANCE.getMedium(buffer);
                    Object value;
                    if (isString) {
                        value = ByteBufferExt.INSTANCE.getString(buffer);
                    } else {
                        value = buffer.getInt();
                    }
                    definition.getParams().put(key, value);
                }
            } else {
                log.warn(String.format("Unhandled definition opcode with id: %s.", opcode));
            }
        } while (true);
        post(definition);
        return definition;
    }

    private static void post(ObjectDefinition definition) {
        if (definition.getWallOrDoor() == -1) {
            definition.setWallOrDoor(0);
            if (definition.getObjectModels() != null && (definition.getObjectTypes() == null || definition.getObjectTypes()[0] == 10)) {
                definition.setWallOrDoor(1);
            }
            if (IntStream.range(0, 5).anyMatch(index -> definition.getActions()[index] != null)) {
                definition.setWallOrDoor(1);
            }
        }
        if (definition.getSupportsItems() == -1) {
            definition.setSupportsItems(definition.getInteractType() != 0 ? 1 : 0);
        }
    }

}

File: app/src/main/kotlin/com/osedit/cache/provider/ObjectProvider.kt
package com.osedit.cache.provider

import com.osedit.cache.util.ByteBufferExt
import com.osedit.spring.domain.Definition
import com.osedit.spring.domain.ItemDefinition
import com.osedit.spring.domain.ObjectDefinition
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer
import java.util.stream.IntStream

object ObjectProvider : Provider<ObjectDefinition> {

    private val log: Logger = LoggerFactory.getLogger(ObjectProvider::class.java)

    override fun decode(buffer: ByteBuffer, definition: ObjectDefinition): Definition {

        do when (val opcode: Int = buffer.get().toInt() and 0xff) {
            1 -> {
                val length: Int = buffer.get().toInt() and 0xff
                when {
                    length > 0 -> {
                        definition.objectTypes = IntArray(length)
                        definition.objectModels = IntArray(length)
                        (0 until length).forEach {
                            definition.objectModels!![it] = buffer.short.toInt() and 0xffff
                            definition.objectTypes!![it] = buffer.get().toInt() and 0xff
                        }
                    }
                }
            }
            2 -> definition.name = ByteBufferExt.getString(buffer)
            5 -> {
                val length: Int = buffer.get().toInt() and 0xff
                when {
                    length > 0 -> {
                        definition.objectTypes = null
                        definition.objectModels = IntStream.range(0, length).map {
                            buffer.short.toInt() and 0xffff
                        }.toArray()
                    }
                }
            }
            14 -> definition.sizeX = buffer.get().toInt() and 0xff
            15 -> definition.sizeY = buffer.get().toInt() and 0xff
            17 -> {
                definition.interactType = 0
                definition.blocksProjectile = false
            }
            18 -> definition.blocksProjectile = false
            19 -> definition.wallOrDoor = buffer.get().toInt() and 0xff
            21 -> definition.contouredGround = 0
            22 -> definition.mergeNormals = true
            23 -> definition.aBool2111 = true
            24 -> {
                definition.animationId = buffer.short.toInt() and 0xffff
                if (definition.animationId == 65535) {
                    definition.animationId = -1
                }
            }
            27 -> definition.interactType = 1
            28 -> definition.decorDisplacement = buffer.get().toInt() and 0xff
            29 -> definition.ambient = buffer.get().toInt()
            39 -> definition.contrast = buffer.get().toInt() * 25
            in 30..34 -> {
                definition.actions[opcode - 30] = ByteBufferExt.getString(buffer)
                if (definition.actions[opcode - 30].equals("Hidden", true)) {
                    definition.actions[opcode - 30] = null
                }
            }
            40 -> {
                val length: Int = buffer.get().toInt() and 0xff
                definition.recolorToFind = ShortArray(length)
                definition.recolorToReplace = ShortArray(length)
                (0 until length).forEach {
                    definition.recolorToFind!![it] = buffer.short
                    definition.recolorToReplace!![it] = buffer.short
                }
            }
            41 -> {
                val length: Int = buffer.get().toInt() and 0xff
                definition.retextureToFind = ShortArray(length)
                definition.retextureToReplace = ShortArray(length)
                (0 until length).forEach {
                    definition.retextureToFind!![it] = buffer.short
                    definition.retextureToReplace!![it] = buffer.short
                }
            }
            62 -> definition.isRotated = true
            64 -> definition.shadow = false
            65 -> definition.modelSizeX = buffer.short.toInt() and 0xffff
            66 -> definition.modelSizeHeight = buffer.short.toInt() and 0xffff
            67 -> definition.modelSizeY = buffer.short.toInt() and 0xffff
            68 -> definition.mapSceneID = buffer.short.toInt() and 0xffff
            69 -> definition.blockingMask = buffer.get().toInt()
            70 -> definition.offsetX = buffer.short.toInt() and 0xffff
            71 -> definition.offsetHeight = buffer.short.toInt() and 0xffff
            72 -> definition.offsetY = buffer.short.toInt() and 0xffff
            73 -> definition.obstructsGround = true
            74 -> definition.isHollow = true
            75 -> definition.supportsItems = buffer.get().toInt() and 0xff
            77 -> {
                var varbitId: Int = buffer.short.toInt() and 0xffff
                if (varbitId == 65535) {
                    varbitId = -1
                }
                definition.varbitId = varbitId
                var varpId: Int = buffer.short.toInt() and 0xffff
                if (varpId == 65535) {
                    varpId = -1
                }
                definition.varpId = varpId
                val length: Int = buffer.get().toInt() and 0xff
                val configChangeDest = IntArray(length + 2)
                IntStream.rangeClosed(0, length).forEach {
                    configChangeDest[it] = buffer.short.toInt() and 0xffff
                    when {
                        configChangeDest[it] == 65535 -> {
                            configChangeDest[it] = -1
                        }
                    }
                }
                configChangeDest[length + 1] = -1
                definition.configChangeDest = configChangeDest
            }
            78 -> {
                definition.ambientSoundId = buffer.short.toInt() and 0xffff
                definition.anInt2083 = buffer.get().toInt() and 0xff
            }
            79 -> {
                definition.anInt2112 = buffer.short.toInt() and 0xffff
                definition.anInt2113 = buffer.short.toInt() and 0xffff
                definition.anInt2083 = buffer.get().toInt() and 0xff
                val length: Int = buffer.get().toInt() and 0xff
                definition.anIntArray2084 = IntStream.range(0, length).map {
                    buffer.short.toInt() and 0xffff
                }.toArray()
            }
            81 -> definition.contouredGround = (buffer.get().toInt() and 0xff) * 256
            82 -> definition.mapAreaId = buffer.short.toInt() and 0xffff
            92 -> {
                var varbitId: Int = buffer.short.toInt() and 0xffff
                if (varbitId == 65535) {
                    varbitId = -1
                }
                definition.varbitId = varbitId
                var varpId: Int = buffer.short.toInt() and 0xffff
                if (varpId == 65535) {
                    varpId = -1
                }
                definition.varpId = varpId
                var value: Int = buffer.short.toInt() and 0xffff
                if (value == 65535) {
                    value = -1
                }
                val length: Int = buffer.get().toInt() and 0xff
                val configChangeDest = IntArray(length + 2)
                IntStream.rangeClosed(0, length).forEach {
                    configChangeDest[it] = buffer.short.toInt() and 0xffff
                    when {
                        configChangeDest[it] == 65535 -> {
                            configChangeDest[it] = -1
                        }
                    }
                }
                configChangeDest[length + 1] = value
                definition.configChangeDest = configChangeDest
            }
            249 -> {
                val length: Int = buffer.get().toInt() and 0xff
                (0 until length).forEach { _ ->
                    val string: Boolean = (buffer.get().toInt() and 0xff) == 1
                    val key: Int = ByteBufferExt.getMedium(buffer)
                    val value: Any = if (string) {
                        ByteBufferExt.getString(buffer)
                    } else {
                        buffer.int
                    }
                    definition.params[key] = value.toString()
                }
            }
            0 -> break
            else -> log.warn(String.format("Unhandled definition opcode with id: %s.", opcode))
        }
        while (true)
        return definition
    }

}
